src/graph/validation.rs

```diff
@@ -0,0 +1,46 @@
+// License: MIT
+// Copyright © 2024 Frequenz Energy-as-a-Service GmbH
+
+//! Methods for validating a [`ComponentGraph`].
+
+mod invariant_checks;
+mod validate_graph;
+mod validate_neighbors;
+
+use crate::{ComponentGraph, Edge, Error, Node};
+
+pub(crate) struct ComponentGraphValidator<'a, N, E>
+where
+    N: Node,
+    E: Edge,
+{
+    cg: &'a ComponentGraph<N, E>,
+    root: &'a N,
+}
+
+pub(crate) fn validate<N, E>(cg: &ComponentGraph<N, E>) -> Result<(), Error>
+where
+    N: Node,
+    E: Edge,
+{
+    let Ok(root) = cg.component(cg.root_id) else {
+        return Err(Error::internal(format!(
+            "Grid component not found with detected component ID: {}.",
+            cg.root_id
+        )));
+    };
+
+    let validator = ComponentGraphValidator { cg, root };
+
+    validator.validate_acyclicity(root, vec![])?;
+    validator.validate_connected_graph(root)?;
+
+    validator.validate_root()?;
+    validator.validate_meters()?;
+    validator.validate_inverters()?;
+    validator.validate_batteries()?;
+    validator.validate_ev_chargers()?;
+    validator.validate_chps()?;
```

Comment:
Not sure how this is normally done in rust, or if there is a popular validation library to use, but when validating, in particular if at some point the validation result needs to be shown to the users, it might be nice to be able to collect all validation errors and show all instead of bailing out on the first error. Maybe for *acyclicity* and *connectivity* it makes sense to bail out because all the other validation assumes a a DAG though.

Reply:
Makes sense. Then you get all the validation issues in one go, instead of finding them out one by one.

Reply:
This is the only pending comment from me but I think it can be perfectly left for a future improvement. Let me know if I should create an issue about it. So LGTM, will leave the approval to a rusty person.

Reply:
Yes, I think this fits well with the config support I'm already working on, so I will do it there.